exporter/latency.go

```diff
@@ -30,9 +39,72 @@ func (e *Exporter) extractLatencyMetrics(ch chan<- prometheus.Metric, c redis.Co
 			var spikeLast, spikeDuration, max int64
 			if _, err := redis.Scan(latencyResult, &eventName, &spikeLast, &spikeDuration, &max); err == nil {
 				spikeDurationSeconds := float64(spikeDuration) / 1e3
-				e.registerConstMetricGauge(ch, "latency_spike_last", float64(spikeLast), eventName)
-				e.registerConstMetricGauge(ch, "latency_spike_duration_seconds", spikeDurationSeconds, eventName)
+				e.registerConstMetricGauge(outChan, "latency_spike_last", float64(spikeLast), eventName)
+				e.registerConstMetricGauge(outChan, "latency_spike_duration_seconds", spikeDurationSeconds, eventName)
 			}
 		}
 	}
 }
+
+func (e *Exporter) extractLatencyHistogramMetrics(outChan chan<- prometheus.Metric, infoAll string, redisConn redis.Conn) {
+	reply, err := redis.Values(doRedisCmd(redisConn, "LATENCY", "HISTOGRAM"))
+	if err != nil {
+		logHistogramErrOnce.Do(func() {
+			log.Errorf("WARNING, LOGGED ONCE ONLY: cmd LATENCY HISTOGRAM, err: %s", err)
+		})
+		log.Debugf("cmd LATENCY HISTOGRAM, err: %s", err)
+		return
+	}
+
+	for i := 0; i < len(reply); i += 2 {
+
+		cmd := string(reply[i].([]byte))
+		details, _ := redis.Values(reply[i+1], nil)
+
+		var totalCalls uint64
+		var bucketInfo []uint64
+
+		if _, err := redis.Scan(details, nil, &totalCalls, nil, &bucketInfo); err != nil {
+			break
+		}
+
+		buckets := map[float64]uint64{}
+
+		for j := 0; j < len(bucketInfo); j += 2 {
+			usec := float64(bucketInfo[j])
+			count := bucketInfo[j+1]
+			buckets[usec] = count
+		}
+
+		totalUsecs := extractTotalUsecForCommand(infoAll, cmd)
+
+		labelValues := []string{"cmd"}
+		e.registerConstHistogram(outChan, "latency_usec", labelValues, totalCalls, float64(totalUsecs), buckets, cmd)
+	}
+}
+
+func extractTotalUsecForCommand(infoAll string, cmd string) uint64 {
+	sanitizedCmd := strings.ReplaceAll(cmd, "|", "\\|")
+	usecRegexp := regexp.MustCompile(`(?m)^cmdstat_` + sanitizedCmd + `(?:\|.*)?:.*usec=([0-9]+).*$`)
+
+	matches := usecRegexp.FindAllStringSubmatch(infoAll, -1)
+
+	if len(matches) == 0 {
+		log.Errorf("Unable to extract total latency for cmd=%s", cmd)
+		return 0
```

Comment:
Did you test this? Can this cause trouble with the histogram when trying to set a total value of zero?

Reply:
Yes, I did test this.

Prometheus will just publish `redis_latency_usec_sum{...} = 0`. This does not seem to affect promql queries using `histogram_quantile()` or similar.

It may be better to log this as a warning rather than an error actually.

Reply:
Ok, cool, thanks for explaining.